### ISSUE TITLE
Listeners Close are async

### ISSUE THREAD

**Issue by NicolasMahe** [mesg-foundation/engine#966]
From @ilgooz comment on PR #895: https://github.com/mesg-foundation/core/pull/895/files/e255920b7c2460127a9bb5805ed4d3b96ef94d7b#diff-4f3848f0a5ec00efd903ac6e8f609a46

- ilgooz 2 days ago
it's better Close() to work in sync. which is the expected behaviour on every close/cleanup. I'll approve this PR but can you please open another pr to improve this on closers?
- krhubert 2 days ago
Can you tell more what do you mean by "work in sync"? to what should be sync close with?

I agree with Ilker, why close are calling a go routine?

**Comment by krhubert** [comment]
doc explains this https://godoc.org/github.com/cskr/pubsub#PubSub.Unsub

**Comment by ilgooz** [comment]
> doc explains this https://godoc.org/github.com/cskr/pubsub#PubSub.Unsub

this is no issue because we already read the messages in an another goroutine. in the current implementation `Unsub()` is not blocked by the reads. please [see](https://github.com/mesg-foundation/engine/pull/895/files/e255920b7c2460127a9bb5805ed4d3b96ef94d7b#diff-604f4192deea5adf9c159e5f88f64630R156)

**Comment by krhubert** [comment]
Yes the messages are read in another goroutine, but docs say:

> Unsub **must** be called from a goroutine that is different from the subscriber

So the sub method is called [here](https://github.com/mesg-foundation/engine/pull/895/files/e255920b7c2460127a9bb5805ed4d3b96ef94d7b#diff-4f3848f0a5ec00efd903ac6e8f609a46R40) this is way unsub is **must** be called in a different goroutine.
I've checked it.

**Comment by ilgooz** [comment]
the doc is a bit tricky. what they mean is you must consume all the data from the data channel before calling the `UnSub()`. otherwise unsub will stuck forever. we already have a listener that consumes data, with a range loop, in a separate goroutine.

anyway, even if this was in the other way around, we should still have closers to work in sync. otherwise there is no way of knowing if everything is shutdown properly before the execution of other code followed by the `Close()`. it's not so important for this case but following the good pattern in general is important.

**Comment by NicolasMahe** [comment]
Let's stop this conversation.
We'll reopen if we have problem with the current implementation.
